phoenix-core/src/main/java/org/apache/phoenix/coprocessor/PhoenixTTLRegionObserver.java

```diff
@@ -0,0 +1,368 @@
+            if (txnScn != null) {
+                TimeRange timeRange = scan.getTimeRange();
+                scan.setTimeRange(timeRange.getMin(), Bytes.toLong(txnScn));
+            }
+            minTimestamp = scan.getTimeRange().getMin();
+            maxTimestamp = scan.getTimeRange().getMax();
+            now = maxTimestamp != HConstants.LATEST_TIMESTAMP ?
+                            maxTimestamp :
+                            EnvironmentEdgeManager.currentTimeMillis();
+        }
+
+        @Override public int getBatch() {
+            return scanner.getBatch();
+        }
+
+        @Override public long getMaxResultSize() {
+            return scanner.getMaxResultSize();
+        }
+
+        @Override public boolean next(List<Cell> result) throws IOException {
+            try {
+                boolean hasMore;
+                do {
+                    hasMore = scanner.next(result);
+                    if (result.isEmpty()) {
+                        break;
+                    }
+                    numRowsScanned++;
+                    if (maskIfExpired && checkRowNotExpired(result)) {
+                        break;
+                    }
+
+                    if (deleteIfExpired && deleteRowIfExpired(result)) {
+                        numRowsDeleted++;
+                        break;
+                    }
+                    // skip this row
+                    // 1. if the row has expired (checkRowNotExpired returned false)
+                    // 2. if the row was not deleted (deleteRowIfExpired returned false and
+                    //  do not want it to count towards the deleted count)
+                    if (maskIfExpired) {
+                        numRowsExpired++;
+                    }
+                    result.clear();
+                } while (hasMore);
+                return hasMore;
+            } catch (Throwable t) {
+                ServerUtil.throwIOException(region.getRegionInfo().getRegionNameAsString(), t);
+                return false; // impossible
+            }
+        }
+
+        @Override public boolean next(List<Cell> result, ScannerContext scannerContext)
+                throws IOException {
+            throw new IOException(
+                    "next with scannerContext should not be called in Phoenix environment");
+        }
+
+        @Override public boolean nextRaw(List<Cell> result, ScannerContext scannerContext)
+                throws IOException {
+            throw new IOException(
+                    "NextRaw with scannerContext should not be called in Phoenix environment");
+        }
+
+        @Override public void close() throws IOException {
+            if (!reported) {
+                LOG.debug(String.format(
+                        "***** PHOENIX-TTL-SCAN-STATS-CLOSE: " + "request-id:[%d] = [%d, %d, %d]",
+                        this.requestId, this.numRowsScanned, this.numRowsExpired,
```

Comment:
So basically, `requestId` is the current count of already masked expired rows? Can you please rename the variable to indicate that and the log line as well?

Reply:
I will change it to be a separate request-id for MASKING and DELETE